### ISSUE TITLE
Non-linear relationships in covariates

### ISSUE THREAD

**Issue by c777s** [WillianFuks/tfcausalimpact#109]
Forgive me as I'm new to CausalImpact and the BSTS modeling on which it's based.

To improve the accuracy of the results, should I split a covariate into multiple tiered covariates to account for non-linear relationships between the covariate and the original time series?

Or does this package already model non-linear relationships in the covariates natively?

**Comment by WillianFuks** [comment]
Hi @c777s ,

The default CausalImpact model does not model nonlinear covariate relationships natively. When covariates are supplied, it adds a
SparseLinearRegression component so the relationship is still linear in the columns provided (`y ≈ time-series component + β₁x₁ + β₂x₂ + ...`)

You can represent nonlinear relationships by adding transformed covariates, such as:

```py
data["x_squared"] = data["x"] ** 2
data["x_log"] = np.log1p(data["x"])
```

Another option is to create spline basis features using a preprocessing package such as scikit-learn’s SplineTransformer
(https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.SplineTransformer.html) or Patsy’s bs()
(https://patsy.readthedocs.io/en/latest/spline-regression.html). For example:

```py
from sklearn.preprocessing import SplineTransformer

transformer = SplineTransformer(
    n_knots=4,
    degree=3,
    include_bias=False,
)

# Fit using only the pre-intervention covariate values to avoid data leakage
transformer.fit(data.iloc[:150][["x"]])

x_spline = transformer.transform(data[["x"]])

spline_data = pd.DataFrame(
    x_spline,
    index=data.index,
    columns=[f"x_spline_{i}" for i in range(x_spline.shape[1])]
)

model_data = pd.concat([data[["y"]], spline_data], axis=1)
ci = CausalImpact(model_data, [0, 149], [150, 199])
```

Tiered or binned covariates can also model piecewise behavior but they require manually choosing thresholds and may overfit, especially with many bins so choose them carefully (only if you identify in your data it could be helpful). Splines usually provide a smoother and more flexible representation.
